adapters/amqp-vertx/src/main/java/org/eclipse/hono/adapter/amqp/AmqpAdapterSaslAuthenticatorFactory.java

```diff
@@ -222,97 +232,86 @@ public void process(final Handler<Boolean> completionHandler) {
                     currentContext.runOnContext(action -> completionHandler.handle(Boolean.TRUE));
                 }
             });
-
-            final byte[] saslResponse = new byte[sasl.pending()];
-            sasl.recv(saslResponse, 0, saslResponse.length);
-
-            if (AuthenticationConstants.MECHANISM_PLAIN.equals(remoteMechanism)) {
-                verifyPlain(saslResponse, deviceAuthTracker);
-            } else if (AuthenticationConstants.MECHANISM_EXTERNAL.equals(remoteMechanism)) {
-                verifyExternal(deviceAuthTracker);
-            }
-
         }
 
         @Override
         public boolean succeeded() {
             return succeeded;
         }
 
-        private void verifyPlain(final byte[] saslResponse, final Handler<AsyncResult<DeviceUser>> completer) {
-
-            currentSpan.log("authenticating device using SASL PLAIN");
+        private Future<String[]> parseSaslResponse(final byte[] saslResponse) {
```

Comment:
why not return `Future<UsernamePasswordCredentials>` instead? IMHO that would be make it easier to understand the semantics of the values returned in the future ...

Reply:
I'd rather keep all the information from the `saslResponse` in the returned object here still (including the `authzid`).

The `VertxBasedAmqpProtocolAdapter` allows for a custom `authenticatorFactory` to be set but still makes quite some assumptions on what the `authenticatorFactory`'s `ProtonSaslAuthenticator` ought to do (e.g. setting certain connection attachments). Therefore I'm thinking towards creating an AuthenticatorFactory base class that provides methods for fulfilling these assumptions while staying rather generic, hence the intention to stay low-level with the sasl response parsing here.

Reply:
I see